### PR TITLE
virts-509-update agent executors on heartbeat

### PR DESCRIPTION
**Pull request by cabanuel**
If agent heartbeat now has changed executors and/or priorities for
executors, handle_heartbeat function now updates the executors in the
database.

### CODE REVIEW

File: app/service/agent_svc.py [hunk 1]
```diff
@@ -72,7 +72,14 @@ def __init__(self, untrusted_timer):
             update_data = dict(last_seen=now, pid=pid, ppid=ppid)
             if agent[0]['trusted']:
                 update_data['last_trusted_seen'] = now
-            await self.data_svc.update('core_agent', 'paw', paw, data=update_data)
+            await self.get_service('data_svc').update('core_agent', 'paw', paw, data=update_data)
+            incoming_executors =[]
+            for i,e in enumerate(executors):
+                incoming_executors.append(dict(executor=e, preferred=1 if i == 0 else 0))
```
Comment (privateducky) [inline app/service/agent_svc.py]: I'd like to see this logic grouped together like the existing create_agent logic so it offers a single responsibility look & feel. how to do this? I would create an update_agent() function here on the data_svc and see if you can fill it in.

good work on the delete portion-- that is easy to overlook but you caught it.

Comment (cabanuel) [reply app/service/agent_svc.py]: I can do that. I was trying to stay away from making modifications in data_svc. I will rework this!

File: app/service/agent_svc.py [hunk 2]
```diff
@@ -72,7 +72,9 @@ def __init__(self, untrusted_timer):
             update_data = dict(last_seen=now, pid=pid, ppid=ppid)
             if agent[0]['trusted']:
                 update_data['last_trusted_seen'] = now
-            await self.data_svc.update('core_agent', 'paw', paw, data=update_data)
+            await self.get_service('data_svc').update('core_agent', 'paw', paw, data=update_data)
+            await self.get_service('data_svc').update_agent('core_executor', dict(agent_id=agent[0]['id'],
```
Comment (wbooth) [inline app/service/agent_svc.py]: self.data_svc vs self.get_service('data_svc')

File: app/service/data_svc.py [hunk 3]
```diff
@@ -390,6 +390,22 @@ def __init__(self, dao):
         :return: None
         """
         await self.dao.update(table, key, value, data)
+        
+    async def update_agent(self, table, data):
+        """
+        Update the agent's core executor
+        :param table:
+        :param data:
+        :return:
```
Comment (wbooth) [inline app/service/data_svc.py]: we should fill these out if we are going to have them

File: app/service/data_svc.py [hunk 4]
```diff
@@ -390,6 +390,22 @@ def __init__(self, dao):
         :return: None
         """
         await self.dao.update(table, key, value, data)
+        
+    async def update_agent(self, table, data):
```
Comment (wbooth) [inline app/service/data_svc.py]: seems this function is very specific to agent executors and may be to say update_agent_executors and drop the table input

**Comment by privateducky** [comment]
waiting on @wbooth's final review

**Comment by cabanuel** [comment]
Does this modified solution encompass that?

`old_executors = [d['executor'] for d in(sorted(previous_executors, key = lambda i: i['preferred'], reverse = True))]

for item in set(new_executors) - set(old_executors):
	await self.dao.create('core_executor', dict(agent_id=agent_id, executor=item, preferred=0))

if old_executors[0] != new_executors[0]:
	await self.dao.delete('core_executor', dict(agent_id=agent_id, executor=old_executors[0]))
	await self.dao.create('core_executor', dict(agent_id=agent_id, executor=old_executors[0]), preferred=0)
	await self.dao.delete('core_executor', dict(agent_id=agent_id, executor=new_executors[0]))
	await self.dao.create('core_executor', dict(agent_id=agent_id, executor=new_executors[0]), preferred=1)

for item in set(old_executors) - set(new_executors):
	await self.dao.delete('core_executor', dict(agent_id=agent_id, executor=item))`

File: app/service/data_svc.py [hunk 5]
```diff
@@ -391,6 +391,28 @@ def __init__(self, dao):
...
+        old_executors = [d['executor'] for d in(sorted(previous_executors, key = lambda i: i['preferred'],
+                                                       reverse = True))]
+
+        for item in set(new_executors) - set(old_executors):
+            await self.dao.create('core_executor', dict(agent_id=agent_id, executor=item, preferred=0))
+
+        if old_executors[0] != new_executors[0]:
+            await self.dao.update('core_executor', 'agent_id', agent_id, data=dict(preferred=0))
+            await self.dao.update('core_executor', 'agent_id', str(agent_id) +
+                                  '" AND executor = "{}'.format(new_executors[0]), data=dict(preferred=1))
```
Comment (wbooth) [inline app/service/data_svc.py]: let's keep all sql construction in the database class.

try adding a unique key on the executors table. then you can do a create here and the preference will be reflected properly.

**Pull request merged**